src/api/python/z3compat/z3.py

```diff
@@ -0,0 +1,255 @@
+    if msg is None:
+        raise Exception("Unimplemented")
+    else:
+        raise Exception("Unimplemented: {}".format(msg))
+
+
+class Z3Exception(Exception):
+    def __init__(self, value):
+        self.value = value
+
+    def __str__(self):
+        return str(self.value)
+
+
+# We use _assert instead of the assert command because we want to
+# produce nice error messages in Z3Py at rise4fun.com
+def _assert(cond, msg):
+    if not cond:
+        raise Z3Exception(msg)
+
+
+# Hack for having nary functions that can receive one argument that is the
+# list of arguments.
+# Use this when function takes a single list of arguments
+def _get_args(args):
+    try:
+        if len(args) == 1 and (isinstance(args[0], tuple) or isinstance(args[0], list)):
+            return list(args[0])
+        else:
+            return list(args)
+    except:  # len is not necessarily defined when args is not a sequence (use reflection?)
+        return list(args)
+
+
+# Use this when function takes multiple arguments
+def _get_args_ast_list(args):
+    try:
+        if isinstance(args, set) or isinstance(args, tuple):
+            return [arg for arg in args]
+        else:
+            return args
+    except:
+        return args
+
+
+def _to_param_value(val):
+    if isinstance(val, bool):
+        if val:
+            return "true"
+        else:
+            return "false"
+    else:
+        return str(val)
+
+
+class Context(object):
+    """A Context manages all terms, configurations, etc.
+
+    In CVC4's API, these are managed by a solver, but we keep the Context class
+    (which just wraps a solver) for compatiblity.
+
+    It's only responsibilities are:
+        * making fresh identifiers for a given sort
+        * looking up previously defined constants
+    """
+
+    def __init__(self, *args, **kws):
+        self.solver = pc.Solver()
+        self.solver.setOption("produce-models", "true")
+        # Map from (name, sort) pairs to constant terms
+        self.vars = {}
+        # An incresing identifier used to make fresh identifiers
+        self.next_fresh_var = 0
+
+    def __del__(self):
+        self.solver = None
+
+    def get_var(self, name, sort):
+        """Get the variable identified by `name`.
+
+        If no variable of that name (with that sort) has been created, creates
+        one.
+
+        Returns a Term
+        """
+        if (name, sort) not in self.vars:
+            self.vars[(name, sort)] = self.solver.mkConst(sort.ast, name)
+        return self.vars[(name, sort)]
+
+    def next_fresh(self, sort, prefix):
+        """Make a name such that (name, sort) is fresh.
+
+        The name will be prefixed by `prefix`"""
+        name = "{}{}".format(prefix, self.next_fresh_var)
+        while (name, sort) in self.vars:
+            self.next_fresh_var += 1
+            name = "{}{}".format(prefix, self.next_fresh_var)
+        return name
+
+    def __eq__(self, o):
+        return self.solver == o.solver
```

Comment:
So `__eq__` is not explicitly defined for `Solver` I don't think. But this probably works because it falls back on object identity, right? It might be clearer to just use that explicitly: `return self.solver is o.solver` if that still works.

Reply:
Done in 8c40f7cbd2

Reply:
Good catch! I honestly missed this.